p2p/transport/webrtc/internal/encoding/hex.go

```diff
@@ -0,0 +1,155 @@
+//
+// Example: { 0x01, 0x02, 0x03 } -> "01:02:03"
+func EncodeInterspersedHexToBuilder(src []byte, builder *strings.Builder) {
+	if len(src) == 0 {
+		return
+	}
+	builder.Grow(len(src)*3 - 1)
+	v := src[0]
+	builder.WriteByte(hextable[v>>4])
+	builder.WriteByte(hextable[v&0x0f])
+	for _, v = range src[1:] {
+		builder.WriteByte(':')
+		builder.WriteByte(hextable[v>>4])
+		builder.WriteByte(hextable[v&0x0f])
+	}
+}
+
+// DecodeInterspersedHex decodes a byte slice string of hex characters into a byte slice,
+// where the hex characters are expected to be separated by a colon (':').
+//
+// Example: {'0', '1', ':', '0', '2', ':', '0', '3'} -> { 0x01, 0x02, 0x03 }
+func DecodeInterspersedHex(src []byte) ([]byte, error) {
+	if len(src) == 0 {
+		return []byte{}, nil
+	}
+	if len(src) < 2 {
+		return nil, hex.ErrLength
+	}
+
+	dst := make([]byte, (len(src)+1)/3)
+	i, j := 0, 1
+	for ; j < len(src); j += 3 { // jump one extra byte for the separator (:)
+		p := src[j-1]
+		q := src[j]
+		if j+1 < len(src) && src[j+1] != ':' {
+			return nil, errUnexpectedInterperseHexChar
+		}
+
+		a := reverseHexTable[p]
+		b := reverseHexTable[q]
+		if a > 0x0f {
+			return nil, hex.InvalidByteError(p)
+		}
+		if b > 0x0f {
+			return nil, hex.InvalidByteError(q)
+		}
+		dst[i] = (a << 4) | b
+		i++
+	}
+	if (len(src)+1)%3 != 0 {
+		if len(src)%3 == 0 {
+			j -= 1
+		}
+		// Check for invalid char before reporting bad length,
+		// since the invalid char (if present) is an earlier problem.
+		if reverseHexTable[src[j-1]] > 0x0f {
+			return nil, hex.InvalidByteError(src[j-1])
+		}
+		return nil, hex.ErrLength
+	}
+	return dst[:i], nil
+}
+
+// DecodeInterpersedHexFromASCIIString decodes an ASCII string of hex characters into a byte slice,
+// where the hex characters are expected to be separated by a colon (':').
+//
+// NOTE that this function returns an error in case the input string contains non-ASCII characters.
+//
+// Example: "01:02:03" -> { 0x01, 0x02, 0x03 }
+func DecodeInterpersedHexFromASCIIString(src string) ([]byte, error) {
+	if len(src) == 0 {
+		return []byte{}, nil
+	}
+	if len(src) < 2 {
+		return nil, hex.ErrLength
+	}
+
+	dst := make([]byte, (len(src)+1)/3)
+	i, j := 0, 1
+	for ; j < len(src); j += 3 { // jump one extra byte for the separator (:)
+		p := src[j-1]
+		q := src[j]
+		if j+1 < len(src) && src[j+1] != ':' {
+			return nil, errUnexpectedInterperseHexChar
+		}
+
+		a := reverseHexTable[p]
+		b := reverseHexTable[q]
+		if a > 0x0f {
+			return nil, hex.InvalidByteError(p)
+		}
+		if b > 0x0f {
+			return nil, hex.InvalidByteError(q)
+		}
+		dst[i] = (a << 4) | b
+		i++
+	}
+	if (len(src)+1)%3 != 0 {
+		if len(src)%3 == 0 {
+			j -= 1
+		}
+		// Check for invalid char before reporting bad length,
+		// since the invalid char (if present) is an earlier problem.
+		if reverseHexTable[src[j-1]] > 0x0f {
+			return nil, hex.InvalidByteError(src[j-1])
+		}
+		return nil, hex.ErrLength
```

Comment:
it's anyway the bad path so no need to optimise for it, but even more so,
I tried to keep this code as close as possible to the original Golang std hex encoding/decoding code.
Just copied it over to allow our interspersing here, so we do not have to do it in two passes.
It's also tested + fuzz tested so it does work, that it can be improved possibly, but not sure that is still required at this point.